### PR TITLE
Add support for multicast ping

### LINKED ISSUE #4: feature request: obtain source address of ICMP reply packet
When doing a broadcast/multicast ping (e.g. to `ip6-allrouters`, i.e. `ff02::2`), I’m interested in the source address of the ICMP reply packet.

Unfortunately, this package does not offer a way to obtain the source address AFAICT. Could a way be added? If you lack the time, I’m happy to contribute a PR, but figured I’d check with you first regarding the preferred API for the feature, provided you’re okay with accepting it.

Thanks in advance!

### PR DESCRIPTION
**Pull request by corny**
@stapelberg Please test and write your opinion. What do you think about the interface?

closes #4

### CODE REVIEW

File: sending.go [hunk 1]
```diff
@@ -30,15 +33,62 @@ func (pinger *Pinger) PingAttempts(remote *net.IPAddr, timeout time.Duration, at
...
-	seq := uint16(atomic.AddUint32(&sequence, 1))
-	req := request{
-		wait: make(chan struct{}),
+	if err != nil {
+		return 0, err
+	}
+	return req.roundTripTime()
+}
+
+// PingMulticast sends a single Echo Request and waits a given time for the answer(s)
```
Comment (stapelberg) [inline sending.go]: Could you explain the channel semantics in the documentation here? I.e., the channel will be closed when the context is cancelled (I think?).

Also, maybe add an example to the documentation?
```go
ctx, canc := context.WithTimeout(context.Background(), 1 * time.Second)
defer canc()
responses, err := PingMulticast(ctx, net.ParseIP("ff02::1%eth0"))
if err != nil {
  return err
}
for resp := range responses {
  fmt.Println(resp)
}
```

**Comment by stapelberg** [comment]
Thanks!

I ran the ping-test tool from the multicast branch and it works fine for my use-case (tried with ff02::1%eth0 and ff02::2%eth0).

I think the interface is fine, and left a comment about a possible documentation improvement.

**Comment by corny** [comment]
Good idea - I will add one example.

Does `PingMulticast(ctx, net.ParseIP("ff02::1%eth0"))` really work? This should be correct:
```
addr, _ := ResolveIPAddr("ip", "ff02::1%eth0")
responses, err := PingMulticast(ctx, addr)
```

**Comment by stapelberg** [comment]
> Does PingMulticast(ctx, net.ParseIP("ff02::1%eth0")) really work? This should be correct:

You are correct, `net.ParseIP()` is not sufficient, but your example does work :).

File: request.go [hunk 2]
```diff
@@ -1,20 +1,42 @@
...
+// A request is a currently running ICMP echo request waiting for multple answers.
+type multiRequest struct {
+	tStart    time.Time // when was the request packet sent?
+	responses chan Response
+	closed    bool
+	mtx       sync.RWMutex
+}
+
+// Response is a response to a multicast echo request
+type Response struct {
```
Comment (dmke) [inline request.go]: A response to an echo request is commonly known as "Reply" → `type Reply`?

File: request.go [hunk 3]
```diff
@@ -33,3 +64,31 @@ func (req *request) roundTripTime() (time.Duration, error) {
...
+func (req *multiRequest) close() {
+	req.mtx.Lock()
+	req.closed = true
+	close(req.responses)
+	req.mtx.Unlock()
+}
+
+// handleResponse is responsible for adding a result to the result set
+func (req *multiRequest) handleResponse(_ error, addr net.IP, tRecv *time.Time) {
+
```
Comment (dmke) [inline request.go]: Remove empty line.

File: sending.go [hunk 4]
```diff
@@ -19,7 +20,9 @@ func (pinger *Pinger) PingAttempts(remote *net.IPAddr, timeout time.Duration, at
 		err = errors.New("zero attempts")
 	} else {
 		for i := 0; i < attempts; i++ {
-			if rtt, err = pinger.Ping(remote, timeout); err == nil {
+			rtt, err = pinger.Ping(remote, timeout)
+
```
Comment (dmke) [inline sending.go]: Remove empty line.

File: sending.go [hunk 5]
```diff
@@ -30,15 +33,66 @@ func (pinger *Pinger) PingAttempts(remote *net.IPAddr, timeout time.Duration, at
...
+	// wait for answer
+	select {
+	case <-req.wait:
+		// already dequeued
+		err = req.result
+	case <-ctx.Done():
+		// dequeue request
+		pinger.removeRequest(seq)
+
```
Comment (dmke) [inline sending.go]: Remove empty line.

File: sending.go [hunk 6]
```diff
@@ -30,15 +33,66 @@ func (pinger *Pinger) PingAttempts(remote *net.IPAddr, timeout time.Duration, at
...
+		return 0, err
+	}
+	return req.roundTripTime()
+}
+
+// PingMulticast sends a single echo request and returns a channel for the responses.
+// The channel will be closed on termination of the context.
+// An error is returned if the sending of the echo request fails.
+func (pinger *Pinger) PingMulticast(ctx context.Context, remote *net.IPAddr) (<-chan Response, error) {
+
```
Comment (dmke) [inline sending.go]: Remove empty line.

File: sending.go [hunk 7]
```diff
@@ -30,15 +33,66 @@ func (pinger *Pinger) PingAttempts(remote *net.IPAddr, timeout time.Duration, at
...
+	req := multiRequest{
+		responses: make(chan Response),
+	}
+	seq, err := pinger.sendRequest(remote, &req)
+
```
Comment (dmke) [inline sending.go]: Remove empty line.

File: request.go (hunk 2 above)
Comment (corny) [reply request.go]: Thanks, has been renamed.

**Comment by corny** [comment]
I renamed `PingMulticast` to `PingMulticastContext` and added a `PingMulticast(destination *net.IPAddr, wait time.Duration)` method. I think passing a duration is more conventient than passing a context.

**Pull request merged**